### PR TITLE
Optional parties and party/signer/role validation overhaul

### PR DESCRIPTION
**Pull request by SpicyLemon**
## Description

closes: #1381
closes: #1437 
closes: #1438 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/provenance-io/provenance/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Wrote unit and integration [tests](https://github.com/provenance-io/provenance/blob/main/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

**Comment by iramiller** [comment]
Going to take a bit to get through this but it looks like a very nice streamlining of the session and party handling systems so far.

File: x/metadata/keeper/signers_utils.go [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+	return p.CanBeUsed() && !p.IsUsed() && p.GetRole() == role
+}
+
+// IsSameAs returns true if this is the same as the provided Party or PartyDetails.
+// Only the address and role are considered for this test.
+func (p *PartyDetails) IsSameAs(p2 types.Partier) bool {
+	return types.SamePartiers(p, p2)
+}
+
+// SignersWrapper is a thing that holds the signers as strings and acc addresses.
```
Comment (Taztingo) [inline x/metadata/keeper/signers_utils.go:190]: Maybe rephrase this as "SignersWrapper stores the signers as strings and acc addresses."

File: x/metadata/keeper/signers_utils.go [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+	return cache
+}
+
+// UnwrapMetadataContext retrieves a Context from a context.Context instance attached with WrapSDKContext.
+// It then adds an AuthzCache to it.
+// It panics if a Context was not properly attached, or if the AuthzCache can't be added.
+//
+// This should be used for all Metadata msg server endpoints instead of sdk.UnwrapSDKContext.
+// This should not be used outside of the Metadata module.
+func UnwrapMetadataContext(goCtx context.Context) sdk.Context {
```
Comment (Taztingo) [inline x/metadata/keeper/signers_utils.go:293]: Neat

File: x/metadata/keeper/signers_utils.go [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+	return findMissingComp(required, toCheck, func(r, c types.Party) bool { return types.SamePartiers(&r, &c) })
+}
+
+// findMissingComp returns all entries in required where an entry does not exist in toCheck
+// such that the provided comp function returns true.
+// Duplicate entries in required do not require duplicate entries in toCheck.
+// E.g. findMissingComp([a, b, a], [a]) => [b], and findMissingComp([a, b, a], [b]) => [a, a].
+func findMissingComp[R any, C any](required []R, toCheck []C, comp func(R, C) bool) []R {
+	var rv []R
+reqLoop:
```
Comment (Taztingo) [inline x/metadata/keeper/signers_utils.go:317]: Same goto comment as above.

**Review by Taztingo — CHANGES_REQUESTED** [review]

File: x/metadata/keeper/signers.go [hunk 4]
```diff
@@ -0,0 +1,541 @@
...
+}
+
+// associateRequiredRoles goes through the required roles, marking parties as used
+// when possible. Returns a list of required role entries that haven't yet been fulfilled.
+//
+// This is similar to validateRolesPresent except this requires a role to have a signer
+// in order for it to fulfill a required role.
+func associateRequiredRoles(parties []*PartyDetails, reqRoles []types.PartyType) []types.PartyType {
+	var missingRoles []types.PartyType
+reqRolesLoop:
```
Comment (Taztingo) [inline x/metadata/keeper/signers.go:119]: Same goto comment as above.

File: x/metadata/keeper/scope.go [hunk 5]
```diff
@@ -277,185 +275,225 @@ func (k Keeper) ValidateScopeUpdate(
...
-		if err != nil {
-			return fmt.Errorf("failed to decode data access address %s : %w", da, err)
-		}
-		found := false
-		for _, pda := range existing.DataAccess {
-			if da == pda {
-				found = true
-				break
+
+dataAccessLoop:
```
Comment (Taztingo) [inline x/metadata/keeper/scope.go:449]: Same goto comment as above.

**Review by Taztingo — CHANGES_REQUESTED** [review]

File: x/metadata/client/cli/cli_test.go [hunk 6]
```diff
@@ -213,13 +213,13 @@ func (s *IntegrationCLITestSuite) SetupSuite() {
...
 	s.recordSpecID = metadatatypes.RecordSpecMetadataAddress(s.contractSpecUUID, s.recordName)
 
-	s.scope = *metadatatypes.NewScope(
-		s.scopeID,
-		s.scopeSpecID,
-		ownerPartyList(s.user1AddrStr),
-		[]string{s.user1AddrStr},
-		s.user2AddrStr,
-	)
+	s.scope = metadatatypes.Scope{
```
[the remark this answers is not in the document]
Comment (SpicyLemon) [reply x/metadata/client/cli/cli_test.go:216]: I'm really not a fan of the `NewScope` function.

1. It doesn't do anything that direct struct instantiation doesn't do.
2. It's less readable since you can't use named parameters when calling functions.
3. If this had already used struct initialization (instead of `NewScope`), I wouldn't have had to update these lines at all.

In all these changes, specially in the unit tests, I favored switching to struct initialization, but I left uses of `NewScope` where switching would have added a bunch of lines.  E.g. In this case, no new lines needed to be added, and switching to struct initialization actually saved an extra line because I didn't need to provide the new `false`. But in some of the other unit tests where the `NewScope` and args are all on one line, I just added the `, false`.

Ideally, it saves us time the next time an argument needs to be added to a scope; we don't have to update tests that continue to pass using the zero-value of the new field.

File: x/metadata/types/scope.go [hunk 7]
```diff
@@ -162,28 +151,34 @@ func (s *Scope) RemoveOwners(addressesToRemove []string) error {
 	if len(addressesToRemove) == 0 {
 		return nil
 	}
+addressesToRemoveLoop:
```
[the remark this answers is not in the document]
Comment (SpicyLemon) [reply x/metadata/types/scope.go:154]: Good call. I'll create a `scope#hasOwnerAddress` func instead of using the label approach.

File: x/metadata/types/scope.go [hunk 8]
```diff
@@ -461,21 +457,38 @@ func (p Party) ValidateBasic() error {
 	return nil
 }
 
+// ValidatePartiesAreUnique makes sure that no two provided parties are the same.
+func ValidatePartiesAreUnique(parties []Party) error {
```
[the remark this answers is not in the document]
Comment (SpicyLemon) [reply x/metadata/types/scope.go:461]: I always have to stop and think about the slice ranges. And I'm used to the i = 0 to 1-n, j = i+1 to n format, so it was easier for me to write it that way and be sure it's not off-by-one somewhere.

A little added reason is that parties are `[]Party`, a slice of concrete (non-reference) values. But `.IsSameAs` needs a reference. The `lParty` and `rParty` variables are re-used for each iteration which can lead to problems if you're storing a reference to them outside the loop. It's not a concern in this case since `.IsSameAs` doesn't hold onto the provided arg, but the `gosec` linter would complain, requiring a nolint comment (as seen in one or two other places).

But really, I feel safer with the old-school for loop iteration than with sub-slices since I'm more familiar with the former when it comes to comparing values in a single list/slice/array.

File: x/metadata/types/scope_test.go [hunk 9]
```diff
@@ -36,90 +36,110 @@ func (s *ScopeTestSuite) SetupSuite() {
 }
 
 func (s *ScopeTestSuite) TestScopeValidateBasic() {
+	ns := func(scopeID, scopeSpecification MetadataAddress, owners []Party, dataAccess []string, valueOwner string) *Scope {
```
[the remark this answers is not in the document]
Comment (SpicyLemon) [reply x/metadata/types/scope_test.go:39]: I wanted the name to be super short like this, and I didn't feel comfortable adding `ns` to the `types` package, even if only for unit tests. By defining it again like that (where multiple tests can use it), it's no different than just using `NewScope`; when it's updated, a lot of things need updating in the exact same way so that the tests can continue to operate as they previously did. By defining it individually for each test that needs it, it limits the amount of changes needed when the scope changes. If a test starts failing, the `ns` function can be updated in just that test as needed, and the other tests (that are still passing) don't require attention.

File: x/metadata/types/scope_test.go [hunk 10]
```diff
@@ -646,6 +696,123 @@ func (s *ScopeTestSuite) TestMetadataAuditUpdate() {
...
+		expErr     string
+	}{
+		{
+			name:       "opt allowed nil parties",
+			optAllowed: true,
+			parties:    nil,
+			expErr:     "",
+		},
+		{
+			name:       "opt allowed empty parties",
```
[the remark this answers is not in the document]
Comment (SpicyLemon) [reply x/metadata/types/scope_test.go:713]: Oops. I put `[]Party{}` in the one below it instead of this one. Good catch.

File: x/metadata/keeper/signers_utils.go (hunk 3 above)
Comment (SpicyLemon) [reply x/metadata/keeper/signers_utils.go:317]: With these loops with labels, I didn't have any need for the just the inner loop functionality (e.g. a `containsComp` function). So, rather than create them and add unit tests on them just so they could be used in the one place I was designing them for, I opted for nesting the loops directly.

Also, I tend to prefer the golang loop labels over the `found` bool way.

And in most of these, extracting the inner loop into its own function, and unit testing that function, doesn't really reduce the unit tests needed on the original function. So it's extra work for something that isn't really needed.

**Review by Taztingo — DISMISSED** [review]

**Review by iramiller — DISMISSED** [review]
High level review of changes complete.  Overall this looks good!

File: x/metadata/types/scope.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (SpicyLemon) [reply x/metadata/types/scope.go:467]: Heh. So, I accidentally checked in a version of `ValidatePartiesAreUnique` that had both the original loops and this:
```golang
	for i, pi := range parties[:len(parties)-1] {
		for j := range parties[i+1:] {
			if pi.IsSameAs(&parties[j]) {
				return fmt.Errorf("duplicate parties not allowed: address = %s, role = %s, indexes: %d, %d",
					parties[i].Address, parties[i].Role, i, j)
			}
		}
	}
```

That new version has some problems that the original didn't:
1. It panics if `parties` is nil.
2. It panics if `parties` is empty.
3. The `j` index is not the same here as it is in the other one. It starts at 0 and goes to len(parties)-1-i. E.g., with three entries, when i = 0, j = 0, 1, then i = 1, j = 0. So that `&parties[j]` is not the correct entry to be comparing.

Basically, it only ever works right when there's exactly 1 entry in `parties`.

Here is what it'd look like after fixing it:
```golang
	if len(parties) <= 1 {
		return nil
	}
	for i, pi := range parties[:len(parties)-1] {
		for jj := range parties[i+1:] {
			j := i + 1 + jj
			if pi.IsSameAs(&parties[j]) {
				return fmt.Errorf("duplicate parties not allowed: address = %s, role = %s, indexes: %d, %d",
					parties[i].Address, parties[i].Role, i, j)
			}
		}
	}
```

I kind of like the original version better though, so I'm going to keep it like that.

Relatedly, I realized that I forgot to add unit tests on it. So I'll add some.

File: x/metadata/client/cli/helpers.go [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+	}
+	_, err = sdk.AccAddressFromBech32(parts[0])
+	if err != nil {
+		return rv, fmt.Errorf("invalid address %q: %w", parts[0], err)
+	}
+	rv.Address = parts[0]
+	if len(parts) >= 2 {
+		// Don't change the role from default unless we know it parsed correctly.
+		rv.Role, err = parsePartyType(parts[1])
+		if err != nil {
```
[the remark this answers is not in the document]
Comment (SpicyLemon) [reply x/metadata/client/cli/helpers.go]: I had to change this back.

That second part needs to only happen when role parsing fails. So I could have it as `if err != nil && len(parts) != 2 { return rv, err }`, but then I'd still need an `if err != nil { ... }` around the second part.

Here's what I had changed it to, and why it didn't work:
```golang
if len(parts) >= 2 {
	// Don't change the role from default unless we know it parsed correctly.
	rv.Role, err = parsePartyType(parts[1])
	if err != nil && len(parts) != 2 {
		return rv, err
	}
	// allow for "<address>,opt"
	var optErr error
	rv.Optional, optErr = parseOptional(parts[1])
	if optErr != nil {
		// If the second (and only) thing wasn't an optional flag either, return the role error.
		return rv, err
	}
}
```

Consider the input string `"cosmos1v9jxgun9wde47v2lta047h6lta047h6lhgvprx,servicer,opt"`
During the parsing of `parts[1]` (`"servicer"`), it correctly parses it into that role and `err` is nil. It then continues on to try to parse "servicer" as an optional, but fails. Since that fails, it then returns the original error, which was nil. That stops it from moving on to try to parse the third part, though.

All that being said, I do like putting that first return closer to where the error is generated. I just still need the second part on level in, under an `if err != nil {...}`.

**Review by iramiller — APPROVED** [review]
With the recent updates all tests/checks are passing.

**Comment by iramiller** [comment]
Massive code coverage now... nice work.
